File: old/sorting/quick_sort/count_total_swaps_in_lamuto_quicksort/count_total_swaps_in_lamuto_quicksort.py
def lamuto_partition(arr, start, end, count):
    left = start
    pivot = arr[end]

    for i in range(start, end):
        if arr[i] < pivot:
            arr[i], arr[left] = arr[left], arr[i]
            left += 1
            count += 1

    arr[end] = arr[left]
    arr[left] = pivot
    count += 1

    return arr, left, count


def lamuto_quick_sort(arr, start, end, count):
    if start >= end:
        return count
    else:
        arr, lamuto_partition_index, count = lamuto_partition(arr, start, end, count)
        count = lamuto_quick_sort(arr, start, lamuto_partition_index - 1, count)
        count = lamuto_quick_sort(arr, lamuto_partition_index + 1, end, count)
        return count


def main():
    input = [3, 2, 1, 4, 5, 6]
    print(input)
    swapCount = lamuto_quick_sort(input, 0, len(input) - 1, 0)
    print(input)
    print(swapCount)


main()

File: old/sorting/quick_sort/count_total_swaps_in_lamuto_quicksort/count_total_swaps_in_lamuto_quicksort.cpp
#include <iostream>
#include <vector>

using namespace std;

int lamutoPartition(vector<int>& arr, int low, int high, int& count) {
    int left = low;
    int pivot = arr[high];

    for (int i = low; i < high; i++) {
        if (arr[i] < pivot) {
            swap(arr[i], arr[left]);
            left++;
            count++;
        }
    }

    arr[high] = arr[left];
    arr[left] = pivot;
    count++;

    return left;
}

void lamutoQuickSort(vector<int>& arr, int low, int high, int& count) {
    if (low >= high) {
        return;
    } 

    int pivotPosAfterPartition = lamutoPartition(arr, low, high, count);
    lamutoQuickSort(arr, low, pivotPosAfterPartition - 1, count);
    lamutoQuickSort(arr, pivotPosAfterPartition + 1, high, count);
}

int main() {
    int swapCount = 0;
    vector<int> input = {3, 2, 1, 4, 5, 6};
    lamutoQuickSort(input, 0, 5, swapCount);
    cout << "swap count: " << swapCount << endl;
    for (int i = 0; i < input.size(); i++) {
        cout << input[i] << " ";
    }
    return 0;
}
